platforms/jvm/plugins-java-base/src/main/java/org/gradle/api/plugins/internal/DefaultJavaPluginExtension.java

```diff
@@ -202,7 +198,7 @@ public void registerFeature(String name, Action<? super FeatureSpec> configureAc
     public void withJavadocJar() {
         maybeEmitMissingJavaComponentDeprecation("withJavadocJar()");
 
-        if (project.getPlugins().hasPlugin(JavaPlugin.class)) {
+        if (isJavaComponentPresent()) {
             JavaPluginHelper.getJavaComponent(project).withJavadocJar();
```

Comment:
We should similarly replace these `JavaPluginHelper.getJavaComponent` to calls like
```
project.getComponents().withType(JvmSoftwareComponentInternal).configureEach(javaComponent -> {
    javaComponent.doTheSameThing()
})
```

Reply:
This is a behavior change in 2 respects:
- It would apply to _any_ `JvmSoftwareComponentInternal` added (I know it's an internal and there _should_ only ever be one)
- It would potentially apply lazily, versus the current eager application.  I'm not sure how this would matter now, since the one component should have already been added, but again, it _could_ differ.

More importantly, if we're going to go with `configureEach`, why bother checking `isJavaComponentPresent` at all?  The fact that we are checking, and even emitting a deprecation warning if it isn't present, makes me think this should be left written as an operation against the single, named component we checked for, rather than a `configureEach` against any component of that type.

These may in practice have the same effect, but without a broader consideration I don't think this changes should be made at this time, as part of this PR.

Reply:
Does this DSL:

```
java {
  withJavadocJar()
}
```

mean enable javadoc publishing on the component named `java`, or enable it on any component of type `JvmSoftwareComponentInternal`?  There (incidentally, at the present time) happens to be only one of latter, and it is named `java`, but I don't think we should encode this assumption here.

Reply:
We can revert these changes and defer the decision to another PR.

In general I think the questions you pose are valid, but in practice:
1. There is only one of these Components ever. Users do not have the ability to create another
2. The component is added eagerly

Given these two points, using `configureEach` will not change behavior.

Reply:
Discussed.  Going to configure each `JvmSoftwareComponentInternal` and not privilege the `java` component.